format/macho/macho.go

```diff
@@ -559,7 +586,71 @@ func machoDecode(d *decode.D, in interface{}) interface{} {
 		}
 		ncmdsIdx++
 	})
-	return nil
+}
+
+func fatParse(d *decode.D) {
+	// Go to start of the file again
+	d.SeekAbs(0)
+	d.FieldStruct("fat_header", func(d *decode.D) {
+		d.FieldRawLen("magic", 8*8)
+		narchs := d.FieldU32("narchs")
+		narchsIdx := 0
+		d.FieldStructArrayLoop("archs", "arch", func() bool {
+			return narchsIdx < int(narchs)
+		}, func(d *decode.D) {
+			// parse FatArch
+			d.FieldStruct("fat_arch", func(d *decode.D) {
+				// beware cputype and cpusubtype changes from ofile header to fat header
+				cpuType := d.FieldU32("cputype", cpuTypes)
+				d.FieldU32("cpusubtype", cpuSubTypes[int64(cpuType)])
+				d.FieldU32("offset")
+				d.FieldU32("size")
+				d.FieldU32("align")
+			})
+		})
+		for i := 0; uint64(i) < narchs; i++ {
+			// parse ofiles
+			ofileDecode(d)
+		}
+	})
+}
+
+func arParse(d *decode.D) {
```

Comment:
Could the existing ar decoder be used for this or is it too macho-specific? It looks like ar but without the `"!<arch>\n"` header? to reuse it i guess it needs some option to skip header and something to hint it how to handle files with certain names to handle SYMDEF and nested ar? maybe too messy

Reply:
Ah there is a header, just noticed the ARMAG thing

Reply:
I totally forgot that we should have an existing ar decoder. Let me actually delete this and re-use the already tested one.

Reply:
deleted the code and re-used ar parser from ar format.

That required main decoder method to be exported as `DecodeAr`. I hope it is fine.

Reply:
You mean exported https://github.com/wader/fq/blob/master/format/ar/ar.go#L27? I try to keep the format implementations decoupled from each other to not cause import cycles and also to force them to have a somewhat clear APIs between each other. So i would prefer if you used the `format.Dependencies` to register that the format needs `fornmat.AR` and then you can use `d.Format*(...)` to actually do the decoding.

Reply:
oh okay. I was not aware of this usage. Will look into its usages.

Reply:
👍  So if you want communicate something to the ar decoder you would create a `format.ArIn` struct and somehow model what you want do, say maybe in this case a map saying that some filenames are suppose to be parser a certain way or that it should try to do nested ar-archives decode etc, but it can be a bit tricky to modell these things nicely. You can also return data by creating a `format.ArOut` struct and return it. But all of this  requires that you do type asserts and such things. 

I have some future plans to allow a user to pass these format specific options directly from jq, ex: `mp4({skip_sample_decode: true})`.